recipes/deepspeed/meta.yaml

```diff
@@ -0,0 +1,96 @@
+{% set name = "deepspeed" %}
+{% set version = "0.7.7" %}
+
+{% set torch_proc_type = "cuda" if cuda_compiler_version != "None" else "cpu" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: 5069fefabbd61dcff64e322189310908ac0291bbda0d472fe89de1cd82d417af
+
+build:
+  number: 0
+  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}  # [cuda_compiler_version != "None"]
+  string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                                # [cuda_compiler_version == "None"]
+  skip: true  # [win or cuda_compiler_version in ("10.2",)]
+
+  # NOTE(hadim): Without this I get `conda_build.exceptions.OverLinkingError` errors such as
+  # `deepspeed/ops/adam/fused_adam_op.cpython-310-x86_64-linux-gnu.so): $RPATH/libc10_cuda.so not found in packages,
+  # sysroot(s) nor the missing_dso_whitelist`
+  missing_dso_whitelist:
+    - '*'
+
+requirements:
+
+  # NOTE(hadim): pytorch is only needed during the build in reqs.build and reqs.host when the ops are being built which
+  # is only possible when CUDA is available. To prevent creating too many builds for pytorch-cpu we remove the pytorch
+  # requirements when CUDA is not available.
+  # WARNING(hadim): What I don't know is whether the below logic will allow conda to pick the CUDA builds during solving.
+
+  build:
+    - python                                           # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}               # [build_platform != target_platform]
+    - pytorch                                          # [(build_platform != target_platform) and cuda_compiler_version not in (undefined, 'None')]
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}                           # [cuda_compiler_version not in (undefined, 'None')]
+  host:
+    - python
+    - pip
+    - git
+    - py-cpuinfo
+    # For the `async_io` op
+    - libaio  # [linux]
+    # Leaving two dependencies helps rerender correctly
+    # The first gets filled in by the global pinnings
+    # The second gets the processor type
+    - pytorch                                          # [cuda_compiler_version not in (undefined, 'None')]
+    - pytorch =*={{ torch_proc_type }}*                # [cuda_compiler_version not in (undefined, 'None')]
+  run:
+    - python
+    - hjson-py
+    # NOTE(hadim): coming at https://github.com/conda-forge/staged-recipes/pull/19098
+    # - ninja-python
```

Comment:
See https://github.com/conda-forge/staged-recipes/pull/19098#issuecomment-1379606877 for the discussion around `ninja`.

Reply:
I'd check to see if this supports system ninja too. If so, the requirement could be just ninja, and it could be patched out.

Reply:
According to https://github.com/microsoft/DeepSpeed/blob/e7c14026d3b7a71e8624e5d15c3c4aae1e7b5bb0/deepspeed/env_report.py#L55, it seems like deepspeed relies on `import ninja`. I understand this is not necessary needed to go that way, but this is how `deepspeed` has been designed for the moment.

Here are the 3 options regarding that recipe:

1. Merge here and people will have to install `ninja-python` from PyPI, or they won't be able to compile ops on-the-fly.
2. Hold and wait for `ninja-python` to be available in conda-forge.
3. Ask the `deepspeed` team to switch to `ninja` without using `ninja-python`.

Assuming this is acceptable, we could execute 1) and then modify the conda package once `ninja-python` is available on conda-forge.

Happy to get other thoughts on this plan.

Reply:
Ping @jeffra here as you are a deepspeed dev in case you have a preference for the above. Please feel free to also ping other's deepspeed devs if relevant.

Reply:
I don't think they should switch; I think the best way is to use it like this:

https://github.com/scikit-build/scikit-build-core/blob/21c4607f543f534bad1a1d7a50462319d8e79345/src/scikit_build_core/program_search.py#L44-L60

(ignore `module`, assume it's `True`)

That way, if a user doesn't have the ninja package installed, the system ninja can still be used.

Reply:
Thanks @henryiii, that looks like the ideal setup indeed.

Reply:
I am going to merge that recipe as it is and open tickets in the feedstock to discuss the ninja dep.

Reply:
xref to https://github.com/conda-forge/deepspeed-feedstock/issues/1 for the record